File: ios/MusicStory/Domain/StoryShareText.swift
import Foundation

enum StoryShareText {
    static func normalizeUserFacingTranscript(_ text: String) -> String {
        text
            .replacingOccurrences(of: "й+утй+уб", with: "YouTube", options: .caseInsensitive)
            .replacingOccurrences(of: "ют+уб", with: "YouTube", options: .caseInsensitive)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func resolveVoicedText(script: String, ttsTranscript: String?) -> String {
        let raw = ttsTranscript?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
            ? ttsTranscript!
            : script
        return normalizeUserFacingTranscript(raw)
    }

    static func resolveVoicedText(_ response: StoryResponse) -> String {
        resolveVoicedText(script: response.script, ttsTranscript: response.ttsTranscript)
    }

    static func excerpt(_ text: String, maxChars: Int = 280) -> String {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.count <= maxChars { return t }
        let cut = String(t.prefix(maxChars))
        if let dot = cut.lastIndex(of: "."), cut.distance(from: cut.startIndex, to: dot) > maxChars / 3 {
            return String(cut[...dot]).trimmingCharacters(in: .whitespacesAndNewlines) + "…"
        }
        if let space = cut.lastIndex(of: " ") {
            return String(cut[..<space]).trimmingCharacters(in: .whitespacesAndNewlines) + "…"
        }
        return cut + "…"
    }

    static func plainShareMessage(artist: String, title: String, voicedText: String) -> String {
        """
        \(title) — \(artist)

        \(voicedText.trimmingCharacters(in: .whitespacesAndNewlines))

        — Эфир AI · https://www.efir-ai.ru
        """
    }

    static func cardVariantSeed(trackKey: String, playedAt: TimeInterval) -> Int {
        let hash = (trackKey + String(Int(playedAt))).hashValue
        return abs(hash) % 4
    }
}

extension StoryHistoryEntry {
    var displayVoicedText: String {
        let v = voicedText?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return v.isEmpty ? script : v
    }
}

File: android/app/src/main/java/com/musicstory/app/domain/StoryShareText.kt
package com.musicstory.app.domain

import com.musicstory.app.data.model.StoryResponse

object StoryShareText {
    fun normalizeUserFacingTranscript(text: String): String =
        text
            .replace(Regex("""й\+?утй\+?уб""", RegexOption.IGNORE_CASE), "YouTube")
            .replace(Regex("""ют\+?уб""", RegexOption.IGNORE_CASE), "YouTube")
            .trim()

    fun resolveVoicedText(script: String, ttsTranscript: String? = null): String {
        val raw = ttsTranscript?.trim()?.takeIf { it.isNotEmpty() } ?: script.trim()
        return normalizeUserFacingTranscript(raw)
    }

    fun resolveVoicedText(response: StoryResponse): String =
        resolveVoicedText(response.script, response.ttsTranscript)

    fun excerpt(text: String, maxChars: Int = 280): String {
        val t = text.trim()
        if (t.length <= maxChars) return t
        val cut = t.take(maxChars)
        val lastSentence = cut.lastIndexOf('.').takeIf { it > maxChars / 3 } ?: cut.lastIndexOf(' ')
        return if (lastSentence > 0) cut.take(lastSentence + 1).trim() + "…" else cut.trim() + "…"
    }

    fun plainShareMessage(artist: String, title: String, voicedText: String): String =
        buildString {
            appendLine("$title — $artist")
            appendLine()
            appendLine(voicedText.trim())
            appendLine()
            append("— Эфир AI · https://www.efir-ai.ru")
        }

    fun cardVariantSeed(trackKey: String, playedAt: Long): Int {
        val hash = (trackKey + playedAt).hashCode()
        return (hash and Int.MAX_VALUE) % 4
    }
}
